### PR TITLE
Use object indirection in HttpContextAccessor (#1066)

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses https://github.com/aspnet/AspNetCore/issues/5144. This is backporting https://github.com/aspnet/AspNetCore/commit/49d785c9343d64b93fadfb7eea36c358825cccb0 to 2.2

### CODE REVIEW

**Review by Tratcher — APPROVED** [review]
Note the branches aren't opening for a few weeks.

**Comment by JunTaoLuo** [comment]
## Description:
In 2.2 we made a change to require request's TraceIdentifier to match when retrieving the HttpContext from the HttpContextAccessor. The goal of the change was to reduce the chance of an invalid HttpContext from being retrieved once the request has ended. However, this caused a regression since if any middleware modifies the TraceIdentifier on the request, the HttpContextAccessor will return a null value for HttpContext. In 3.0 a different approach is used to mitigate the retrieval of invalid HttpContext without relying on the TraceIdentifier. As a side-effect, the regression in behaviour was also removed.

## Customer Impact
As mentioned in #5144, any customers on 2.2 will not be able to retrieve HttpContext from HttpContextAccessor if they modify the TraceIdentifier of the request.

## Regression?
Regression from 2.1.

## Risk:
Low. The fix has already been made to 3.0, we are just backporting it to 2.2

**Review by davidfowl — APPROVED** [review]
We should add a regression test for that case we missed.

**Comment by vivmishra** [comment]
Approved for 2.2.2

**Comment by dougbu** [comment]
Before merging, please update the 2.2.2 section of eng/PatchConfig.props: https://github.com/aspnet/Extensions/blob/d8407116d183debaadb801c0cbc41d65927999d6/eng/PatchConfig.props#L7-L10

File: src/Http/Http/src/HttpContextAccessor.cs [hunk 1]
```diff
@@ -7,21 +7,35 @@ namespace Microsoft.AspNetCore.Http
...
-                // Only return the context if the stored request id matches the stored trace identifier
-                // context.TraceIdentifier is cleared by HttpContextFactory.Dispose.
-                return value.traceIdentifier == value.context?.TraceIdentifier ? value.context : null;
+                return  _httpContextCurrent.Value?.Context;
             }
             set
             {
-                _httpContextCurrent.Value = (value?.TraceIdentifier, value);
+                var holder = _httpContextCurrent.Value;
+                if (holder != null)
```
Comment (keithjac) [inline src/Http/Http/src/HttpContextAccessor.cs:21]: Why is this here? What does 'trapped in the AsyncLocals' mean, as this holder is completely private? Who else could have it?

Doesn't this introduce a short window where the getter will return null while you're in the middle of setting a new value, and why would you want that?
